src/graphics/vulkan.zig

```diff
@@ -0,0 +1,150 @@
+	VK_ERROR_FRAGMENTATION = -1000161000,
+	VK_ERROR_INVALID_OPAQUE_CAPTURE_ADDRESS = -1000257000,
+	VK_PIPELINE_COMPILE_REQUIRED = 1000297000,
+	VK_ERROR_NOT_PERMITTED = -1000174001,
+	VK_ERROR_SURFACE_LOST_KHR = -1000000000,
+	VK_ERROR_NATIVE_WINDOW_IN_USE_KHR = -1000000001,
+	VK_SUBOPTIMAL_KHR = 1000001003,
+	VK_ERROR_OUT_OF_DATE_KHR = -1000001004,
+	VK_ERROR_INCOMPATIBLE_DISPLAY_KHR = -1000003001,
+	VK_ERROR_VALIDATION_FAILED_EXT = -1000011001,
+	VK_ERROR_INVALID_SHADER_NV = -1000012000,
+	VK_ERROR_IMAGE_USAGE_NOT_SUPPORTED_KHR = -1000023000,
+	VK_ERROR_VIDEO_PICTURE_LAYOUT_NOT_SUPPORTED_KHR = -1000023001,
+	VK_ERROR_VIDEO_PROFILE_OPERATION_NOT_SUPPORTED_KHR = -1000023002,
+	VK_ERROR_VIDEO_PROFILE_FORMAT_NOT_SUPPORTED_KHR = -1000023003,
+	VK_ERROR_VIDEO_PROFILE_CODEC_NOT_SUPPORTED_KHR = -1000023004,
+	VK_ERROR_VIDEO_STD_VERSION_NOT_SUPPORTED_KHR = -1000023005,
+	VK_ERROR_INVALID_DRM_FORMAT_MODIFIER_PLANE_LAYOUT_EXT = -1000158000,
+	VK_ERROR_FULL_SCREEN_EXCLUSIVE_MODE_LOST_EXT = -1000255000,
+	VK_THREAD_IDLE_KHR = 1000268000,
+	VK_THREAD_DONE_KHR = 1000268001,
+	VK_OPERATION_DEFERRED_KHR = 1000268002,
+	VK_OPERATION_NOT_DEFERRED_KHR = 1000268003,
+	VK_ERROR_INVALID_VIDEO_STD_PARAMETERS_KHR = -1000299000,
+	VK_ERROR_COMPRESSION_EXHAUSTED_EXT = -1000338000,
+	VK_PIPELINE_BINARY_MISSING_KHR = 1000483000,
+	VK_ERROR_NOT_ENOUGH_SPACE_KHR = -1000483000,
+	VK_ERROR_INCOMPATIBLE_SHADER_BINARY_EXT = 1000482000,
+	VK_RESULT_MAX_ENUM = 2147483647,
+};
+
+pub fn checkResult(result: c.VkResult) void {
+	const resultEnum = std.meta.intToEnum(VkResultEnum, result) catch {
+		std.log.err("Encountered a vulkan error with unknown error code {}", .{result});
+		return;
+	};
+	if(resultEnum == .VK_SUCCESS) return;
+	std.log.err("Encountered a vulkan error: {s}", .{@tagName(resultEnum)});
+}
+
+fn checkResultIfAvailable(result: anytype) void {
+	if(@TypeOf(result) != void) {
+		checkResult(result);
+	}
+}
+
+fn allocEnumerationGeneric(function: anytype, allocator: NeverFailingAllocator, args: anytype) []@typeInfo(@typeInfo(@TypeOf(function)).@"fn".params[@typeInfo(@TypeOf(function)).@"fn".params.len - 1].type.?).pointer.child {
+	const T = @typeInfo(@typeInfo(@TypeOf(function)).@"fn".params[@typeInfo(@TypeOf(function)).@"fn".params.len - 1].type.?).pointer.child;
+	var count: u32 = 0;
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, null}));
+	const list = allocator.alloc(T, count);
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, list.ptr}));
+	return list;
+}
+
+pub fn enumerateInstanceLayerProperties(allocator: NeverFailingAllocator) []c.VkLayerProperties {
+	return allocEnumerationGeneric(c.vkEnumerateInstanceLayerProperties, allocator, .{});
+}
+
+pub fn enumerateInstanceExtensionProperties(allocator: NeverFailingAllocator, layerName: ?[*:0]u8) []c.VkExtensionProperties {
+	return allocEnumerationGeneric(c.vkEnumerateInstanceExtensionProperties, allocator, .{layerName});
+}
+
+pub const Instance = struct { // MARK: Instance
```

Comment:
Are you sure that mirroring Vulkan nomenclature and structure makes sense in our context?
We only have one instance which represents our application, we are highly unlikely to make use of more than one GPU, well, that where possible simplifications end I think.

Reply:
Yeah, it could make sense to group the devices and instance into one namespace, I guess.